dnsmasq_interface.c

```diff
@@ -464,32 +470,35 @@ void FTL_reply(unsigned short flags, char *name, struct all_addr *addr, int id)
 		// Hereby, this query is now fully determined
 		queries[i].complete = true;
 	}
-	else if(flags & F_FORWARD)
+	else if((flags & F_FORWARD) && isExactMatch)
 	{
-		int domainID = queries[i].domainID;
-		validate_access("domains", domainID, true, __LINE__, __FUNCTION__, __FILE__);
+		// Save reply type and update individual reply counters
+		save_reply_type(flags, i, response);
 
-		if(strcmp(domains[domainID].domain, name) == 0)
+		// If received NXDOMAIN and AD bit is set, Quad9 may have blocked this query
+		if(flags & F_NXDOMAIN && queries[i].AD)
 		{
-			// Save reply type and update individual reply counters
-			save_reply_type(flags, i, response);
-
-			// If received NXDOMAIN and AD bit is set, Quad9 may have blocked this query
-			if(flags & F_NXDOMAIN && queries[i].AD)
-			{
-				query_externally_blocked(i);
-			}
-
-			// Detect if returned IP indicates that this query was blocked
-			detect_blocked_IP(flags, answer, i);
+			query_externally_blocked(i);
 		}
+
+		// Detect if returned IP indicates that this query was blocked
+		detect_blocked_IP(flags, answer, i);
 	}
 	else if(flags & F_REVERSE)
 	{
+		// isExactMatch is not used here as the PTR is special.
+		// Example:
+		// Question: PTR 8.8.8.8
+		// will lead to:
+		//   domains[domainID].domain = 8.8.8.8.in-addr.arpa
+		// and will return
+		//   name = google-public-dns-a.google.com
+		// Hence, isExactMatch is always false
+
 		// Save reply type and update individual reply counters
 		save_reply_type(flags, i, response);
 	}
-	else
+	else if(isExactMatch && !queries[i].complete)
```

Comment:
Made the `else` clause a bit more specific as there are several acceptable things that fail to match any `if` above.